source/unix/openssl_rsa.c

```diff
@@ -0,0 +1,397 @@
+    if (ctx == NULL) {
+        return aws_raise_error(AWS_ERROR_SYS_CALL_FAILURE);
+    }
+
+    if (EVP_PKEY_encrypt_init(ctx) <= 0) {
+        aws_raise_error(AWS_ERROR_SYS_CALL_FAILURE);
+        goto on_error;
+    }
+
+    if (s_set_enc_ctx_from_algo(ctx, algorithm)) {
+        goto on_error;
+    }
+
+    size_t ct_len = out->capacity - out->len;
+    if (EVP_PKEY_encrypt(ctx, out->buffer, &ct_len, plaintext.ptr, plaintext.len) <= 0) {
+        aws_raise_error(AWS_ERROR_SYS_CALL_FAILURE);
+        goto on_error;
+    }
+    out->len += ct_len;
+
+    EVP_PKEY_CTX_free(ctx);
+
+    return AWS_OP_SUCCESS;
+
+on_error:
+    EVP_PKEY_CTX_free(ctx);
+    return AWS_OP_ERR;
+}
+
+int s_rsa_decrypt(
+    struct aws_rsa_key_pair *key_pair,
+    enum aws_rsa_encryption_algorithm algorithm,
+    struct aws_byte_cursor ciphertext,
+    struct aws_byte_buf *out) {
+    struct lc_rsa_key_pair *key_pair_impl = key_pair->impl;
+
+    EVP_PKEY_CTX *ctx = EVP_PKEY_CTX_new(key_pair_impl->key, NULL);
+    if (ctx == NULL) {
+        return aws_raise_error(AWS_ERROR_SYS_CALL_FAILURE);
+    }
+
+    if (EVP_PKEY_decrypt_init(ctx) <= 0) {
+        aws_raise_error(AWS_ERROR_SYS_CALL_FAILURE);
+        goto on_error;
+    }
+
+    if (s_set_enc_ctx_from_algo(ctx, algorithm)) {
+        goto on_error;
+    }
+
+    size_t ct_len = out->capacity - out->len;
+    if (EVP_PKEY_decrypt(ctx, out->buffer, &ct_len, ciphertext.ptr, ciphertext.len) <= 0) {
+        aws_raise_error(AWS_ERROR_SYS_CALL_FAILURE);
+        goto on_error;
+    }
+    out->len += ct_len;
+
+    EVP_PKEY_CTX_free(ctx);
+
+    return AWS_OP_SUCCESS;
+
+on_error:
+    EVP_PKEY_CTX_free(ctx);
+    return AWS_OP_ERR;
+}
+
+int s_set_sign_ctx_from_algo(EVP_PKEY_CTX *ctx, enum aws_rsa_signing_algorithm algorithm) {
+    if (algorithm == AWS_CAL_RSA_SIGNATURE_PKCS1_5_SHA256) {
+        if (EVP_PKEY_CTX_set_rsa_padding(ctx, RSA_PKCS1_PADDING) <= 0) {
+            return aws_raise_error(AWS_ERROR_SYS_CALL_FAILURE);
+        }
+        if (EVP_PKEY_CTX_set_signature_md(ctx, EVP_sha256()) <= 0) {
+            return aws_raise_error(AWS_ERROR_SYS_CALL_FAILURE);
+        }
+    } else if (algorithm == AWS_CAL_RSA_SIGNATURE_PSS_SHA256) {
+        if (EVP_PKEY_CTX_set_rsa_padding(ctx, RSA_PKCS1_PSS_PADDING) <= 0) {
+            return aws_raise_error(AWS_ERROR_SYS_CALL_FAILURE);
+        }
+
+        if (EVP_PKEY_CTX_set_signature_md(ctx, EVP_sha256()) <= 0) {
+            return aws_raise_error(AWS_ERROR_SYS_CALL_FAILURE);
+        }
+    } else {
+        return aws_raise_error(AWS_ERROR_CAL_UNSUPPORTED_ALGORITHM);
+    }
+
+    return AWS_OP_SUCCESS;
+}
+
+int s_rsa_sign(
+    struct aws_rsa_key_pair *key_pair,
+    enum aws_rsa_signing_algorithm algorithm,
+    struct aws_byte_cursor digest,
+    struct aws_byte_buf *out) {
+    struct lc_rsa_key_pair *key_pair_impl = key_pair->impl;
+
+    EVP_PKEY_CTX *ctx = EVP_PKEY_CTX_new(key_pair_impl->key, NULL);
```

Comment:
I had a couple initial concerns with pulling contexts into constructor. We need 4 context (encrypt, decrypt, sign, verify), which seemed like overkill to init when the common flow would be to load the key, perform one operation and destroy key. It also complicates error flow a bit, since for private key all 4 contexts are valid, but for public key only 2 are valid. My last concern is with state of EVP_PKEY_CTX (padding) and thread safety. CTX is not thread safe and modifying padding kills any cross thread usage. Doing something similar to windows with thread once will work here. 

all of those are minor and can be worked around, but rsa is not in a super critical path and going for simpler code seemed like a better choice. let me play around with it a bit to see if there is an easy way to reuse CTX's